spec/domain/repository/persist_repository_from_upstream_spec.rb

```diff
@@ -0,0 +1,243 @@
+# frozen_string_literal: true
+
+require "rails_helper"
+
+RSpec.describe Repository::PersistRepositoryFromUpstream do
+  describe "#create_or_update_from_host_data" do
+    let(:raw_upstream_data) { build(:raw_upstream_data) }
+
+    it "creates a new Repository" do
+      expect { described_class.create_or_update_from_host_data(raw_upstream_data) }.to change(Repository, :count).from(0).to(1)
+    end
+
+    it "assigns the correct data to the Repository" do
+      described_class.create_or_update_from_host_data(raw_upstream_data)
+
+      expected_attributes = described_class.raw_data_repository_attrs(raw_upstream_data)
+
+      expect(Repository.first.attributes.symbolize_keys).to include(**expected_attributes)
+    end
+
+    context "with existing repository" do
+      let!(:existing_repository) { create(:repository, full_name: raw_upstream_data.full_name) }
+
+      it "does not create a new Repository" do
+        expect { described_class.create_or_update_from_host_data(raw_upstream_data) }.not_to change(Repository, :count)
```

Comment:
This should probably check that the ID is also the same, otherwise the method could delete the existing one and replace it with a new one.